### PR TITLE
feat: add disable and enable methods to Mockiavelli class (#9)

### PR DESCRIPTION
**Pull request by Fiszcz**

### CODE REVIEW

File: src/controllers/BrowserControllerFactory.ts [hunk 1]
```diff
@@ -1,5 +1,5 @@
 import * as puppeteer from 'puppeteer';
-import playwright from 'playwright-core';
+import playwright from 'playwright-chromium';
```
Comment (lukaszfiszer) [inline src/controllers/BrowserControllerFactory.ts:2]: why this change?

File: test/integration/mockiavelli.int.test.ts [hunk 2]
```diff
@@ -706,4 +706,20 @@ describe.each(CONTROLLERS)(`Mockiavelli integration (%s)`, (controller) => {
         });
         await mock.waitForRequestsCount(2);
     });
+
+    test('disable() method disables mocking of requests', async () => {
+        const mockedFun = jest.fn().mockReturnValue({ status: 200 });
+        await ctx.mockiavelli.mockGET('/example', mockedFun);
+
+        await ctx.makeRequest('GET', '/example');
+        expect(mockedFun).toHaveBeenCalledTimes(1);
```
Comment (lukaszfiszer) [inline test/integration/mockiavelli.int.test.ts:715]: instead of relying on mock, use public API for assertions like in other tests. For example, by asserting on `response.status`,

**Review by lukaszfiszer — CHANGES_REQUESTED** [review]
Looks good, but please see comments in files

File: src/controllers/BrowserControllerFactory.ts (hunk 1 above)
Comment (Fiszcz) [reply src/controllers/BrowserControllerFactory.ts:2]: Because I have upgraded **playwright-chromium** to the newest version.
And in the new **playwright-chromium** package there are other dependencies (without **playwright-core** package).
And **playwright-chromium** has the same typescript types as **playwright-core**.

File: test/integration/mockiavelli.int.test.ts (hunk 2 above)
Comment (Fiszcz) [reply test/integration/mockiavelli.int.test.ts]: Ok, changed :)

File: src/controllers/PlaywrightController.ts [hunk 3]
```diff
@@ -1,25 +1,36 @@
...
 
 export class PlaywrightController implements BrowserController {
+    private requestHandler?: (request: playwright.Route) => void;
+
     constructor(private readonly page: playwright.Page) {}
 
     async startInterception(onRequest: BrowserRequestHandler) {
-        this.page.route('**/*', (route: playwright.Route) => {
+        await this.stopInterception();
+        this.requestHandler = (route: playwright.Route) => {
```
[the remark this answers is not in the document]
Comment (Fiszcz) [reply src/controllers/PlaywrightController.ts:19]: Ok. And what do you think about this version:
```ts
    private onRequest: BrowserRequestHandler = () => {};
    private hasInterceptingEnabled: boolean = false;

    constructor(private readonly page: playwright.Page) {}

    requestHandler = (route: playwright.Route) => {
        this.onRequest(
            this.toBrowserRequest(route.request()),
            (data) => this.respond(route, data),
            () => this.skip(route)
        );
    };

    async startInterception(onRequest: BrowserRequestHandler) {
        if (this.hasInterceptingEnabled === false) {
            this.hasInterceptingEnabled = true;
            this.onRequest = onRequest;
            await this.page.route('**/*', this.requestHandler);
        }
    }

    async stopInterception() {
        if (this.hasInterceptingEnabled) {
            await this.page.unroute('**/*', this.requestHandler);
            this.hasInterceptingEnabled = false;
        }
    }
```
?

and what if onRequest will change (theoretically) ?

File: README.md [hunk 4]
```diff
@@ -290,6 +290,28 @@ mockiavelli.mockGET('http://api.example.com/api/users', { status: 200 });
 // GET http://another-domain.example.com/api/users => 404
 ```
 
+### Stop mocking
```
Comment (Fiszcz) [inline README.md:293]: Add anchor and add this section to summary.

File: src/controllers/PlaywrightController.ts (hunk 3 above)
[the remark this answers is not in the document]
Comment (lukaszfiszer) [reply src/controllers/PlaywrightController.ts:19]: Now I see where the need of `this.requestHandler` was coming from - we needed somehow capture the reference to `onRequest` callback.

As you said, onRequest changes only theoretically - in practice it does change at all, because it is always passed`Mockiavelli.onRequest` method . So simplify things even further it could be passed to the constructor (through BrowserControllerFactory.createForPage). Then start and stop interception methods would has no args.

[the remark this answers is not in the document]
Comment (Fiszcz) [reply src/controllers/PlaywrightController.ts:19]: So, what do you think about that one:
```typescript
    constructor(private readonly page: playwright.Page, private readonly onRequest: BrowserRequestHandler = () => {}) {}

    requestHandler = (route: playwright.Route) => {
        this.onRequest(
            this.toBrowserRequest(route.request()),
            (data) => this.respond(route, data),
            () => this.skip(route)
        );
    };

    async startInterception() {
           await this.page.route('**/*', this.requestHandler);
    }

    async stopInterception() {
            await this.page.unroute('**/*', this.requestHandler);
        }
    }
```
?

[the remark this answers is not in the document]
Comment (lukaszfiszer) [reply src/controllers/PlaywrightController.ts:19]: 👍

**Comment by lukaszfiszer** [comment]
Merged in separated PR #28
